File: test/evidence/engines.mjs
// JavaScript's answers for cases.tsv: as rule engines usually compile (flags `gi`), and on folded
// text with ASCII semantics (`i`, never `u`). Prints: name <TAB> raw <TAB> folded.
import { readFileSync } from "node:fs";
import { fileURLToPath } from "node:url";
import { dirname, join } from "node:path";

const INVISIBLE = /[­᠎​-‏‪-‮⁠-⁤﻿]/;
const DASH = /[‐-―−﹘﹣－]/;

// NFKC one code point at a time, invisible characters out, dashes to "-".
export function fold(text) {
  let out = "";
  for (const character of text) {
    for (const piece of character.normalize("NFKC")) {
      if (INVISIBLE.test(piece)) continue;
      out += DASH.test(piece) ? "-" : piece === " " || piece === " " ? "\n" : piece === " " ? " " : piece;
    }
  }
  return out;
}

const unescape = (s) => s.replace(/\\u([0-9a-fA-F]{4})/g, (_, hex) => String.fromCharCode(parseInt(hex, 16)));
const yes = (b) => (b ? "yes" : "no");

const file = join(dirname(fileURLToPath(import.meta.url)), "cases.tsv");
for (const line of readFileSync(file, "utf8").split("\n")) {
  if (!line || line.startsWith("#")) continue;
  const [name, pattern, escaped] = line.split("\t");
  const input = unescape(escaped);
  const raw = new RegExp(pattern, "gi").test(input);
  const folded = new RegExp(pattern, "i").test(fold(input));
  console.log([name, yes(raw), yes(folded)].join("\t"));
}

File: test/evidence/engines.py
"""Python's answers for cases.tsv.

As rule engines usually compile (IGNORECASE | UNICODE), and on folded text with ASCII semantics
(IGNORECASE | ASCII). Prints: name <TAB> raw <TAB> folded.
"""

import re
import unicodedata
from pathlib import Path

INVISIBLE = re.compile("[­᠎​-‏‪-‮⁠-⁤﻿]")
DASH = re.compile("[‐-―−﹘﹣－]")


def fold(text: str) -> str:
    """NFKC one code point at a time, invisible characters out, dashes to "-"."""
    out = []
    for character in text:
        for piece in unicodedata.normalize("NFKC", character):
            if INVISIBLE.match(piece):
                continue
            if DASH.match(piece):
                piece = "-"
            elif piece in "  ":
                piece = "\n"
            elif piece == " ":
                piece = " "
            out.append(piece)
    return "".join(out)


def unescape(text: str) -> str:
    return re.sub(r"\\u([0-9a-fA-F]{4})", lambda m: chr(int(m.group(1), 16)), text)


def yes(value) -> str:
    return "yes" if value else "no"


for line in (Path(__file__).parent / "cases.tsv").read_text(encoding="utf-8").splitlines():
    if not line or line.startswith("#"):
        continue
    name, pattern, escaped = line.split("\t")
    text = unescape(escaped)
    raw = re.search(pattern, text, re.IGNORECASE | re.UNICODE)
    folded = re.search(pattern, fold(text), re.IGNORECASE | re.ASCII)
    print("\t".join([name, yes(raw), yes(folded)]))
